File: modified_bsearch/ceiling_of_a_num.rb
# frozen_string_literal: false

require 'minitest/autorun'
require 'minitest/pride'

puts <<~STATEMENT
  Given an array of numbers sorted in an ascending
  order, find the ceiling of a given number ‘key’.
  The ceiling of the ‘key’ will be the smallest
  element in the given array greater than or equal
  to the ‘key’.

  Write a function to return the index of the ceiling
  of the ‘key’. If there isn’t any ceiling return -1.
STATEMENT

# Solution
class Solution
  def solution(nums, key)
    nums.bsearch_index { |num| num >= key } || -1
  end
end

# Tests
class SolutionTest < Minitest::Test
  def test_case1
    nums = [4, 6, 10]
    solution = Solution.new.solution(nums, 6)
    expected = 1

    assert_equal(expected, solution)
  end

  def test_case2
    nums = [1, 3, 8, 10, 15]
    solution = Solution.new.solution(nums, 12)
    expected = 4

    assert_equal(expected, solution)
  end

  def test_case3
    nums = [4, 6, 10]
    solution = Solution.new.solution(nums, 17)
    expected = -1

    assert_equal(expected, solution)
  end

  def test_case4
    nums = [4, 6, 10]
    solution = Solution.new.solution(nums, -1)
    expected = 0

    assert_equal(expected, solution)
  end

  def test_case5
    nums = [1]
    solution = Solution.new.solution(nums, 1)
    expected = 0

    assert_equal(expected, solution)
  end

  def test_case6
    nums = [1]
    solution = Solution.new.solution(nums, 99)
    expected = -1

    assert_equal(expected, solution)
  end
end

File: modified_bsearch/ceiling_of_a_num.js
/*
    Statement:
    Given an array of numbers sorted in an ascending
    order, find the ceiling of a given number ‘key’.
    The ceiling of the ‘key’ will be the smallest
    element in the given array greater than or equal
    to the ‘key’.

    Write a function to return the index of the ceiling
    of the ‘key’. If there isn’t any ceiling return -1.
*/

const ceilingOfANum = (nums, key) => {
    let left = 0;
    let right = nums.length - 1;

    while (left < right) {
        let mid = Math.floor((left + right) / 2);

        if (nums[mid] >= key) right = mid;
        else left = mid + 1;
    }
    return nums[left] >= key ? left : -1;
}

module.exports = ceilingOfANum;
